src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs

```diff
@@ -403,6 +407,31 @@ public static Vector256<ushort> Load256(ref byte ptr)
                 (Vector128<ushort> lower, Vector128<ushort> upper) = Vector128.Widen(Vector128.LoadUnsafe(ref ptr));
                 return Vector256.Create(lower, upper);
             }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            [CompExactlyDependsOn(typeof(Avx))]
+            public static bool Compare256(ref byte left, ref ushort right)
+            {
+                Debug.Assert(Vector256<byte>.Count == Vector256<ushort>.Count * 2);
+
+                Vector256<byte> leftNotWidened = Vector256.LoadUnsafe(ref left);
+                if (!AllCharsInVectorAreAscii(leftNotWidened))
+                {
+                    return false;
+                }
+
+                (Vector256<ushort> lower, Vector256<ushort> upper) = Vector256.Widen(leftNotWidened);
+                Vector256<ushort> rightValues0 = Vector256.LoadUnsafe(ref right);
+                Vector256<ushort> rightValues1 = Vector256.LoadUnsafe(ref Unsafe.Add(ref right, (uint)Vector256<ushort>.Count));
+
+                if (!Vector256<ushort>.AllBitsSet.Equals(
+                    Vector256.BitwiseAnd(Vector256.Equals(lower, rightValues0), Vector256.Equals(upper, rightValues1))))
```

Comment:
We should prefer the operators where possible:
```suggestion
                if (Vector256<ushort>.AllBitsSet != (Vector256.Equals(lower, rightValues0) & Vector256.Equals(upper, rightValues1)))
```

`bool Equals()` in particular is the same as `==` for integral types, but isn't directly recognized as intrinsic and is an instance methjod. So the JIT has to inline and elide the `reference taken` for `this`.

Reply:
Might be nice to just make this `return Vector256<ushort>.AllBitsSet == (Vector256.Equals(lower, rightValues0) & Vector256.Equals(upper, rightValues1))` instead as well

Reply:
Look at the "Side-note on weird codegen observed" section at the bottom of the PR description.

Reply:
I think it should be `==` if you noticed a suboptimal codegen - we'd better look and fix it in JIT instead of complicating C# code just to squeeze everything here and now

Reply:
It should be as simple as:
```cs
return ((lower ^ rightValues0) | (upper ^ rightValues1)) == Vector256.Zero;
```

Reply:
E.g.:

![image](https://github.com/dotnet/runtime/assets/523221/15d453ce-5ece-4fc6-af30-ff812eaa4594)

Reply:
Ooh, that does look nice. The removal of the if statement is still problematic code-gen wise. Here is a simple example showing the difference
https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwAYCwAUDgIxFEBuEATgATUCmKtAvLQGIAUJuAlANxFGzdG3bZufQcRIBOTsIHlCwAPaqANh06wEtAB68iAbyK1ztKADNanfbQCktHrQCEbPmYunCFv7UwAdlorCA0UBmk/AF8vcyDaBGoAV0iiWMIiNU0OCV0DI0IfPwT7Jxd3WlxpaKA=

Reply:
Yep, sometimes it can be fixed with `return cond ? true : false;`

Reply:
> ```cs
> return ((lower ^ rightValues0) | (upper ^ rightValues1)) == Vector256.Zero;
> ```

This should result in better codegen for ARM64 which does not have a `PMOVMSKB` equivalent.

See also: [Bit twiddling with Arm Neon: beating SSE movemasks, counting bits and more](https://community.arm.com/arm-community-blogs/b/infrastructure-solutions-blog/posts/porting-x86-vector-bitmask-optimizations-to-arm-neon).

Reply:
How about:

```csharp
Vector256<ushort> equals = Vector256.Equals(lower, rightValues0) & Vector256.Equals(upper, rightValues1);
if (equals.AsByte().ExtractMostSignificantBits() != 0xffffffffu)
    return false;
return true;
```

Reply:
> How about:

Not needed, `((lower ^ rightValues0) | (upper ^ rightValues1)) == Vector256.Zero;` is just a canonical way to do that. Also, `ExtractMostSignificantBits` is very expensive on arm

Reply:
@EgorBo Agreed, `VPMOVMSKB` has no equivalent on ARM64 (see https://github.com/dotnet/runtime/pull/87141#issuecomment-1583043973), but this method is guarded by `[CompExactlyDependsOn(typeof(Avx))]`.

Reply:
Slighter better codegen for the below:

```csharp
Vector256<ushort> notEquals = (lower ^ rightValues0) | (upper ^ rightValues1);
if (notEquals.AsByte().ExtractMostSignificantBits() != 0)
    return false;
return true;
```

Reply:
> @EgorBo Agreed, `VPMOVMSKB` has no equivalent on ARM64 (see [#87141 (comment)](https://github.com/dotnet/runtime/pull/87141#issuecomment-1583043973)), but this method is guarded by `[CompExactlyDependsOn(typeof(Avx))]`.

still, `== Vector.Zero` is expected to be lowered to MoveMask with SSE2 or to vptest with SSE41/AVX so no reason to do it by hands